elements/lisk-codec/src/collection.ts

```diff
@@ -95,7 +118,155 @@ export const writeObject = (
+	let index = offset;
+	const result: GenericObject = {};
+	for (let i = 0; i < compiledSchema.length; i += 1) {
+		const typeSchema = compiledSchema[i];
+		if (Array.isArray(typeSchema)) {
+			// Takeout the root wireType and field number
+			if (typeSchema[0].schemaProp.type === 'array') {
+				const [arr, nextOffset] = readArray(message, index, typeSchema);
+				result[typeSchema[0].propertyName] = arr;
+				index = nextOffset;
+			} else if (typeSchema[0].schemaProp.type === 'object') {
+				// It should be wire type 2 as it's object
+				const [, keySize] = readUInt32(message, index);
+				index += keySize;
+				// Takeout the length
+				const [, objectSize] = readUInt32(message, index);
+				index += objectSize;
+				const [obj, nextOffset] = readObject(message, index, typeSchema);
+				result[typeSchema[0].propertyName] = obj;
+				index = nextOffset;
+			} else {
+				throw new Error('Invalid container type');
+			}
+			continue;
+		}
+		if (typeSchema.schemaProp.type === 'object' || typeSchema.schemaProp.type === 'array') {
+			// typeSchema is header, and we ignroe this
+			continue;
+		}
+		// Takeout the root wireType and field number
+		const [key, keySize] = readUInt32(message, index);
+		const [fieldNumber] = readKey(key);
+		if (fieldNumber !== typeSchema.schemaProp.fieldNumber) {
+			continue;
+		}
+		// Index is only incremented when the key is actually used
+		index += keySize;
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+		const [scalarValue, scalarSize] = _readers[typeSchema.schemaProp.dataType as string](message, index);
+		index += scalarSize;
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+		result[typeSchema.propertyName] = scalarValue;
+	}
+	return [result, index];
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const readArray = (message: Buffer, offset: number, compiledSchema: CompiledSchemasArray): [Array<any>, number] => {
+	// Takeout the root wireType and field number
+	let index = offset;
+	if (index >= message.length) {
+		return [[], index];
+	}
+	const startingByte = message[index];
+	// It should be wire type 2 as it's object
+	const [, keySize] = readUInt32(message, index);
+	index += keySize;
+	// Takeout the length
+	const [arrayLength, wireType2Size] = readUInt32(message, index);
+	index += wireType2Size;
+	const [, typeSchema] = compiledSchema;
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	const result: Array<any> = [];
+	// Case of object
+	if (Array.isArray(typeSchema)) {
+		// handle as object
+		const [nestedTypeSchema] = typeSchema;
+		if (nestedTypeSchema.schemaProp.type === 'object') {
+			// Since first key and length is read, we do read them directly
+			if (wireType2Size !== 0) {
+				// readObject returns Next offset, not index used
+				const [res, nextOffset] = readObject(message, index, typeSchema);
+				result.push(res);
```

Comment:
Not very important but at `result.push(res)`, either `result` or `res` could be renamed for better understanding